### ISSUE TITLE
Calling Enumerable.ToArray on an Array throws NotSupprtedException in linked app

### ISSUE THREAD

**Issue by eerhardt** [mono/linker#1398]
Using the following csproj and code

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <PublishTrimmed>true</PublishTrimmed>
    <TrimMode>link</TrimMode>
  </PropertyGroup>

  <Target Name="EnsureAllAssembliesAreLinked" AfterTargets="PrepareForILLink"> 
    <ItemGroup>
      <ManagedAssemblyToLink>
        <TrimMode>link</TrimMode>
      </ManagedAssemblyToLink>
    </ItemGroup>
  </Target>
</Project>
```

```C#
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

static class Program
{
    static int Main(string[] args)
    {
        Thing[] parameters = new Thing[]
        {
            new Thing() { Name = "One"},
            new Thing() { Name = "Two"},
        };

        var collection = parameters.ToReadOnly();

        Console.WriteLine("Done");
        return 100;
    }

    public static ReadOnlyCollection<T> ToReadOnly<T>(this IEnumerable<T> enumerable)
    {
        T[] array = enumerable.ToArray<T>();
        return null;
    }

    class Thing
    {
        public string Name { get; set; }
    }
}
```

0. Using the latest .NET 5 SDK
1. `dotnet publish -r win-x64`
2. Run the app `bin\Debug\net5.0\win-x64\publish\ReproArrayBug.exe`

You get exception:

```
Unhandled exception. System.NotSupportedException: Collection was of a fixed size.
   at System.SZArrayHelper.Add[T](T value)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Program.ToReadOnly[T](IEnumerable`1 enumerable) in F:\DotNetTest\ReproArrayBug\Program.cs:line 27
   at Program.Main(String[] args) in F:\DotNetTest\ReproArrayBug\Program.cs:line 13                     
```

Note: Using a local build of master, I've also seen the exception as:

```
Unhandled exception. System.NotSupportedException: Collection is read-only.
   at System.SZArrayHelper.Clear[T]()
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Program.ToReadOnly[T](IEnumerable`1 enumerable) in F:\git\runtime2\artifacts\bin\trimmingTests\projects\Microsoft.Extensions.DependencyInjection.TrimmingTests\ReproArrayBug\win-x64\ReproArrayBug.cs:line 27
   at Program.Main(String[] args) in F:\git\runtime2\artifacts\bin\trimmingTests\projects\Microsoft.Extensions.DependencyInjection.TrimmingTests\ReproArrayBug\win-x64\ReproArrayBug.cs:line 19
```

cc @MichalStrehovsky @vitek-karas @marek-safar

**Comment by MichalStrehovsky** [comment]
I'll take a look. It looks like the runtime is dispatching to the wrong implementation of `ICollection<T>.CopyTo` for arrays. Linker probably messed up the way the secret knock works.

(The `System.SZArrayHelpers` class in CoreLib has a bunch of generic methods on it that implement all the generic interfaces on array types - this is where calls like `ICollection<T>.CopyTo` on an array type actually land.)

**Comment by MichalStrehovsky** [comment]
Fix in dotnet/runtime#40260.
